src/util/requests/WsConnection.cpp

```diff
@@ -0,0 +1,180 @@
+{
+    headers_.insert(headers_.end(), std::make_move_iterator(headers.begin()), std::make_move_iterator(headers.end()));
+    return *this;
+}
+
+WsConnectionBuilder&
+WsConnectionBuilder::setTarget(std::string target)
+{
+    target_ = std::move(target);
+    return *this;
+}
+
+WsConnectionBuilder&
+WsConnectionBuilder::setTimeout(std::chrono::milliseconds timeout)
+{
+    timeout_ = timeout;
+    return *this;
+}
+
+WsConnectionBuilder&
+WsConnectionBuilder::setSslEnabled(bool sslEnabled)
+{
+    sslEnabled_ = sslEnabled;
+    return *this;
+}
+
+Expected<WsConnectionPtr, RequestError>
+WsConnectionBuilder::connect(asio::yield_context yield) const
+{
+    if (sslEnabled_) {
+        auto streamData = impl::SslWsStreamData::create(yield);
+        if (not streamData.has_value())
+            return Unexpected{std::move(streamData.error())};
+
+        if (!SSL_set_tlsext_host_name(streamData->stream.next_layer().native_handle(), host_.c_str())) {
+            beast::error_code errorCode;
+            errorCode.assign(static_cast<int>(::ERR_get_error()), beast::net::error::get_ssl_category());
+            return Unexpected{RequestError{"SSL setup failed", errorCode}};
+        }
+        return connectImpl(std::move(streamData.value()), yield);
+    }
+
+    return connectImpl(impl::WsStreamData{yield}, yield);
+}
+
+template <class StreamDataType>
+Expected<WsConnectionPtr, RequestError>
+WsConnectionBuilder::connectImpl(StreamDataType&& streamData, asio::yield_context yield) const
+{
+    auto context = asio::get_associated_executor(yield);
+    beast::error_code errorCode;
+
+    // These objects perform our I/O
+    asio::ip::tcp::resolver resolver(context);
+    auto& ws = streamData.stream;
+
+    // Look up the domain name
+    auto const results = resolver.async_resolve(host_, port_, yield[errorCode]);
+    if (errorCode)
+        return Unexpected{RequestError{"Resolve error", errorCode}};
+
+    // Set a timeout on the operation
+    beast::get_lowest_layer(ws).expires_after(timeout_);
+
+    // Make the connection on the IP address we get from a lookup
+    auto endpoint = beast::get_lowest_layer(ws).async_connect(results, yield[errorCode]);
+    if (errorCode)
+        return Unexpected{RequestError{"Connect error", errorCode}};
+
+    if constexpr (StreamDataType::sslEnabled) {
+        beast::get_lowest_layer(ws).expires_never();
+        // Perform the SSL handshake
+        ws.next_layer().async_handshake(asio::ssl::stream_base::client, yield[errorCode]);
+        if (errorCode)
+            return Unexpected{RequestError{"SSL handshake error", errorCode}};
+    }
+
+    // Turn off the timeout on the tcp_stream, because
+    // the websocket stream has its own timeout system.
+    beast::get_lowest_layer(ws).expires_never();
+
+    // Set suggested timeout settings for the websocket
+    ws.set_option(websocket::stream_base::timeout::suggested(beast::role_type::client));
```

Comment:
looks like the timeout_ is timeout for connect only? here we set the timeout to suggested?

Reply:
Yeah, probably I should change the name of the method setting timeout_. Suggested timeout for client role is:
- 30 seconds for handshake
- none for idle
- false for keep alive pings (as I understand server should sends pings and boost replies automatically)
I think it looks fine.